text/0000-network-spec.md

```diff
@@ -68,42 +86,116 @@ When one peer creates outgoing connection to another peer, they send a `Handshak
+
+### Localization
+
+We maintain a graph with all connections from our point of view. We should build a DAG/Tree (this should be similar to have several routes for some peer) in such a way that every path from us to other nodes is:
+
+- reasonably short.
+- use high reputation nodes.
+
+One idea to localize nodes in the network:
+
+- From our POV we have infinite reputation.
+- Build a directed graph where weight of edge [u -> v] is reputation of node [u].
+- Build a maximum weighted directed spanning tree.
+
+We only care for parts of the graph such that:
+
+- We should be connected right now because we belong to the same shard (same for block producer).
+- We are going to become part of the same shard soon. (same for block producer).
+- They are nodes with high reputation from our POV.
+
+Localization will be useful to determine whether a node/edge is relevant to us and relevant to a connected peer (it cares about). We will only send and accept new connections of relevant node/edges. If some connected peer spam us with irrelevant connections we should take action towards it. This action can be:
+
+- decrease reputation
+- ban
+- notify + expect better behavior + ban otherwise
+
+## Communication
+
+Broadcasted messges:
 
-- 
+- *Create connection between two nodes*. We should only broadcast based on [localization]. If a new connection is relevant to us we broadcast to other nodes as an important hop. Note: We can force a maximum depth, so we don't send a connection if its exceed some depth. If we connect to other node this is relevant to us. (This way we keep track of all peer at distance 2 from us).
+
+- *Remove connection between two nodes*. We should not broadcast a removal if we didn't broadcast the creation. If node A receives a Remove connection [A, B] signed by B we should disconnect from B (probably malicious action by B). Also it is expected that nodes disconnect from each other when rebalancing network on epoch changes.
+
+Broadcasting messages at the network level is the main source of network overhead unrelated to blockchain protocol level. We should keep this as low as possible.
+
+### Syncing graphs
+
+When we connect to a new peer we should exchange relevant portions of the network to each other progressively. Portions that involve only high reputable nodes and validators.
+
+Edges are signed from both parties with timestamp. They should expire automatically after some period of time So they are re-broadcasted again. This period of time should be reasonably large so incur in small network overhead penalty.
+
+### Routing
+
+Routing mechanism specification.
+
+- *Routing to*: A peer can only send a message to another peer (target) if it knows a path to it in its current view of the network. It will send a message through shortest path using Round Robin (among all direct peers that have shortest path to target it will pick one).
```

Comment:
how we going to do retry in cases of lost message / malicious actor in the middle / broken connection?

Reply:
Should we handle this at the network level. Aren't we assuming at the protocol level that messages can get lost / dropped. What we can try is to identify at which position the issue take place and try to:
- ban or 
- remove edge and update network view.